Description: fix: mark client-rendered damage for backing first-frame gating
 During the window-map-to-first-client-frame window, the composite backing
 pixmap of an override-redirect window holds a CopyArea residue that must
 not be shown. Mark damage whose source is client rendering (clientMark
 set in damageDamageBox and Present) and flip _DDE_NET_WM_BACKING_PIXMAP_STATE
 from 0 to 1 when such content first reaches a Manual backing; the
 temporary Manual sensor is then unregistered.
Author: UnionTech <support@uniontech.com>
Forwarded: not-needed
Last-Update: 2026-09-09
---
Index: xorg-server/composite/compalloc.c
===================================================================
--- xorg-server.orig/composite/compalloc.c
+++ xorg-server/composite/compalloc.c
@@ -46,6 +46,7 @@
 #endif
 
 #include "compint.h"
+#include "damagestr.h"
 
 static Bool
 compScreenUpdate(ClientPtr pClient, void *closure)
@@ -74,6 +75,93 @@ compMarkAncestors(WindowPtr pWin)
 }
 
 static void
+compRefreshRegion(WindowPtr pWin, RegionPtr reg)
+{
+    ScreenPtr pScreen = pWin->drawable.pScreen;
+    WindowPtr pParent = pWin->parent;
+    CompWindowPtr cw = GetCompWindow(pWin);
+    PixmapPtr pPixmap;
+    int srcx, srcy, nBox;
+    BoxPtr pBox;
+
+    if (!pParent || pWin->redirectDraw == RedirectDrawNone || !cw)
+        return;
+    pPixmap = (*pScreen->GetWindowPixmap) (pWin);
+    if (!pPixmap)
+        return;
+    nBox = RegionNumRects(reg);
+    if (!nBox)
+        return;
+    pBox = RegionRects(reg);
+    srcx = pPixmap->screen_x - pParent->drawable.x;
+    srcy = pPixmap->screen_y - pParent->drawable.y;
+
+    cw->refreshing = TRUE;
+    if (pParent->drawable.depth == pWin->drawable.depth) {
+        GCPtr pGC = GetScratchGC(pWin->drawable.depth, pScreen);
+
+        if (pGC) {
+            ChangeGCVal val;
+
+            val.val = IncludeInferiors;
+            ChangeGC(NullClient, pGC, GCSubwindowMode, &val);
+            ValidateGC(&pPixmap->drawable, pGC);
+            while (nBox--) {
+                (*pGC->ops->CopyArea) (&pParent->drawable,
+                                       &pPixmap->drawable,
+                                       pGC,
+                                       pBox->x1 + srcx, pBox->y1 + srcy,
+                                       pBox->x2 - pBox->x1, pBox->y2 - pBox->y1,
+                                       pBox->x1, pBox->y1);
+                pBox++;
+            }
+            FreeScratchGC(pGC);
+        }
+    }
+    else {
+        PictFormatPtr pSrcFormat = PictureWindowFormat(pParent);
+        PictFormatPtr pDstFormat = PictureWindowFormat(pWin);
+        XID inferiors = IncludeInferiors;
+        int error;
+        PicturePtr pSrcPicture = CreatePicture(None, &pParent->drawable, pSrcFormat,
+                                               CPSubwindowMode, &inferiors,
+                                               serverClient, &error);
+        PicturePtr pDstPicture = CreatePicture(None, &pPixmap->drawable, pDstFormat,
+                                               0, 0, serverClient, &error);
+
+        if (pSrcPicture && pDstPicture) {
+            while (nBox--) {
+                CompositePicture(PictOpSrc, pSrcPicture, NULL, pDstPicture,
+                                 pBox->x1 + srcx, pBox->y1 + srcy, 0, 0,
+                                 pBox->x1, pBox->y1,
+                                 pBox->x2 - pBox->x1, pBox->y2 - pBox->y1);
+                pBox++;
+            }
+        }
+        if (pSrcPicture)
+            FreePicture(pSrcPicture, 0);
+        if (pDstPicture)
+            FreePicture(pDstPicture, 0);
+    }
+    cw->refreshing = FALSE;
+}
+
+void
+compRefreshWindowBacking(WindowPtr pWin)
+{
+    RegionRec full;
+    BoxRec box;
+
+    box.x1 = 0;
+    box.y1 = 0;
+    box.x2 = pWin->drawable.width;
+    box.y2 = pWin->drawable.height;
+    RegionInit(&full, &box, 1);
+    compRefreshRegion(pWin, &full);
+    RegionUninit(&full);
+}
+
+static void
 compReportDamage(DamagePtr pDamage, RegionPtr pRegion, void *closure)
 {
     WindowPtr pWin = (WindowPtr) closure;
@@ -81,13 +169,41 @@ compReportDamage(DamagePtr pDamage, Regi
     CompScreenPtr cs = GetCompScreen(pScreen);
     CompWindowPtr cw = GetCompWindow(pWin);
 
-    if (!cs->pendingScreenUpdate) {
-        QueueWorkProc(compScreenUpdate, serverClient, pScreen);
-        cs->pendingScreenUpdate = TRUE;
+    if (cw->update == CompositeRedirectAutomatic) {
+        if (!cs->pendingScreenUpdate) {
+            QueueWorkProc(compScreenUpdate, serverClient, pScreen);
+            cs->pendingScreenUpdate = TRUE;
+        }
+        cw->damaged = TRUE;
+        compMarkAncestors(pWin);
+    }
+    else if (cw->update == CompositeRedirectManual && pDamage->clientMark &&
+             !cw->refreshing) {
+        BoxRec box;
+        RegionRec full, missing;
+
+        box.x1 = 0;
+        box.y1 = 0;
+        box.x2 = pWin->drawable.width;
+        box.y2 = pWin->drawable.height;
+        RegionInit(&full, &box, 1);
+        RegionUnion(&cw->clientDamage, &cw->clientDamage, pRegion);
+        RegionNull(&missing);
+        RegionSubtract(&missing, &full, &cw->clientDamage);
+        if (RegionNotEmpty(&missing)) {
+            compRefreshRegion(pWin, &missing);
+        }
+        else {
+            cw->clientFull = TRUE;
+            if (cw->damageRegistered) {
+                DamageUnregister(cw->damage);
+                cw->damageRegistered = FALSE;
+            }
+        }
+        RegionUninit(&missing);
+        RegionUninit(&full);
     }
-    cw->damaged = TRUE;
-
-    compMarkAncestors(pWin);
+    DamageEmpty(pDamage);
 }
 
 static void
@@ -99,6 +215,7 @@ compDestroyDamage(DamagePtr pDamage, voi
     cw->damage = 0;
     cw->damaged = 0;
     cw->damageRegistered = 0;
+    RegionUninit(&cw->clientDamage);
 }
 
 static Bool
@@ -195,6 +312,9 @@ compRedirectWindow(ClientPtr pClient, Wi
         cw->oldy = COMP_ORIGIN_INVALID;
         cw->damageRegistered = FALSE;
         cw->damaged = FALSE;
+        RegionNull(&cw->clientDamage);
+        cw->clientFull = FALSE;
+        cw->refreshing = FALSE;
         cw->pOldPixmap = NullPixmap;
         dixSetPrivate(&pWin->devPrivates, CompWindowPrivateKey, cw);
     }
@@ -206,10 +326,6 @@ compRedirectWindow(ClientPtr pClient, Wi
         if (!anyMarked)
             anyMarked = compMarkWindows(pWin, &pLayerWin);
 
-        if (cw->damageRegistered) {
-            DamageUnregister(cw->damage);
-            cw->damageRegistered = FALSE;
-        }
         cw->update = CompositeRedirectManual;
     }
     else if (cw->update == CompositeRedirectAutomatic && !cw->damageRegistered) {
@@ -618,13 +734,15 @@ compAllocPixmap(WindowPtr pWin)
         pWin->redirectDraw = RedirectDrawManual;
 
     compSetPixmap(pWin, pPixmap, bw);
+    RegionEmpty(&cw->clientDamage);
+    cw->clientFull = FALSE;
+    cw->refreshing = FALSE;
     cw->oldx = COMP_ORIGIN_INVALID;
     cw->oldy = COMP_ORIGIN_INVALID;
     cw->damageRegistered = FALSE;
-    if (cw->update == CompositeRedirectAutomatic) {
-        DamageRegister(&pWin->drawable, cw->damage);
-        cw->damageRegistered = TRUE;
-    }
+
+    DamageRegister(&pWin->drawable, cw->damage);
+    cw->damageRegistered = TRUE;
     status = TRUE;
 
 out:
Index: xorg-server/composite/compint.h
===================================================================
--- xorg-server.orig/composite/compint.h
+++ xorg-server/composite/compint.h
@@ -97,8 +97,13 @@ typedef struct _CompWindow {
     int oldy;
     PixmapPtr pOldPixmap;
     int borderClipX, borderClipY;
+    RegionRec clientDamage;
+    Bool clientFull;
+    Bool refreshing;
 } CompWindowRec, *CompWindowPtr;
 
+extern void compRefreshWindowBacking(WindowPtr pWin);
+
 #define COMP_ORIGIN_INVALID	    0x80000000
 
 typedef struct _CompSubwindows {
Index: xorg-server/miext/damage/damage.c
===================================================================
--- xorg-server.orig/miext/damage/damage.c
+++ xorg-server/miext/damage/damage.c
@@ -303,6 +303,26 @@ damageRegionProcessPending(DrawablePtr p
 
 }
 
+void
+DamageSetClientMark(DrawablePtr pDrawable)
+{
+    DamagePtr pDmg;
+
+    drawableDamage(pDrawable);
+    for (pDmg = pDamage; pDmg; pDmg = pDmg->pNext)
+        pDmg->clientMark = TRUE;
+}
+
+static void
+_damageSetClientMarks(DrawablePtr pDrawable, Bool mark)
+{
+    DamagePtr pDmg;
+
+    drawableDamage(pDrawable);
+    for (pDmg = pDamage; pDmg; pDmg = pDmg->pNext)
+        pDmg->clientMark = mark;
+}
+
 #if DAMAGE_DEBUG_ENABLE
 #define damageDamageBox(d,b,m) _damageDamageBox(d,b,m,__FUNCTION__)
 static void
@@ -315,6 +335,7 @@ damageDamageBox(DrawablePtr pDrawable, B
 {
     RegionRec region;
 
+    _damageSetClientMarks(pDrawable, TRUE);
     RegionInit(&region, pBox, 1);
 #if DAMAGE_DEBUG_ENABLE
     _damageRegionAppend(pDrawable, &region, TRUE, subWindowMode, where);
@@ -322,6 +343,7 @@ damageDamageBox(DrawablePtr pDrawable, B
     damageRegionAppend(pDrawable, &region, TRUE, subWindowMode);
 #endif
     RegionUninit(&region);
+    _damageSetClientMarks(pDrawable, FALSE);
 }
 
 static void damageValidateGC(GCPtr, unsigned long, DrawablePtr);
Index: xorg-server/miext/damage/damage.h
===================================================================
--- xorg-server.orig/miext/damage/damage.h
+++ xorg-server/miext/damage/damage.h
@@ -110,6 +110,9 @@ extern _X_EXPORT void
 extern _X_EXPORT void
  DamageSetReportAfterOp(DamagePtr pDamage, Bool reportAfter);
 
+extern _X_EXPORT void
+ DamageSetClientMark(DrawablePtr pDrawable);
+
 extern _X_EXPORT DamageScreenFuncsPtr DamageGetScreenFuncs(ScreenPtr);
 
 #endif                          /* _DAMAGE_H_ */
Index: xorg-server/miext/damage/damagestr.h
===================================================================
--- xorg-server.orig/miext/damage/damagestr.h
+++ xorg-server/miext/damage/damagestr.h
@@ -47,6 +47,7 @@ typedef struct _damage {
     DamageDestroyFunc damageDestroy;
 
     Bool reportAfter;
+    Bool clientMark;
     RegionRec pendingDamage;    /* will be flushed post submission at the latest */
     ScreenPtr pScreen;
 } DamageRec;
Index: xorg-server/present/present_scmd.c
===================================================================
--- xorg-server.orig/present/present_scmd.c
+++ xorg-server/present/present_scmd.c
@@ -598,6 +598,7 @@ present_execute(present_vblank_ptr vblan
                 } else
                     damage = &window->clipList;
 
+                DamageSetClientMark(&vblank->window->drawable);
                 DamageDamageRegion(&vblank->window->drawable, damage);
                 return;
             }
Index: xorg-server/composite/compwindow.c
===================================================================
--- xorg-server.orig/composite/compwindow.c
+++ xorg-server/composite/compwindow.c
@@ -400,6 +400,18 @@ compMoveWindow(WindowPtr pWin, int x, in
     pScreen->MoveWindow = compMoveWindow;
 
     compFreeOldPixmap(pWin);
+    {
+        CompWindowPtr cw = GetCompWindow(pWin);
+
+        if (cw && pWin->redirectDraw != RedirectDrawNone && !cw->clientFull) {
+            RegionEmpty(&cw->clientDamage);
+            if (!cw->damageRegistered) {
+                DamageRegister(&pWin->drawable, cw->damage);
+                cw->damageRegistered = TRUE;
+            }
+            compRefreshWindowBacking(pWin);
+        }
+    }
     compCheckTree(pScreen);
 }
 
@@ -416,6 +428,18 @@ compResizeWindow(WindowPtr pWin, int x,
     pScreen->ResizeWindow = compResizeWindow;
 
     compFreeOldPixmap(pWin);
+    {
+        CompWindowPtr cw = GetCompWindow(pWin);
+
+        if (cw && pWin->redirectDraw != RedirectDrawNone && !cw->clientFull) {
+            RegionEmpty(&cw->clientDamage);
+            if (!cw->damageRegistered) {
+                DamageRegister(&pWin->drawable, cw->damage);
+                cw->damageRegistered = TRUE;
+            }
+            compRefreshWindowBacking(pWin);
+        }
+    }
     compCheckTree(pWin->drawable.pScreen);
 }
 
